Discuss *parameterized* vs. *prepared* statements (#38)
ossf/secure-sw-dev-fundamentals: secure_software_development_fundamentals.md

Even if the database language is not SQL, if it is an attack on a language for a database system it is often called an SQL injection attack (even though this is technically not accurate). We will focus on SQL, because SQL is very common and once you understand how to counter SQL injection attacks, it is easy to generalize this to any database language.

> Here is a trivial example; here is a snippet of Java that tries to do an SQL query, but does it insecurely::

Reviewer: ```suggestion
Here is a trivial example; here is a snippet of Java that tries to do an SQL query, but does it insecurely:
```

~~~~java

>     String QueryString = "select * from authors where lastname = ' " + search_lastname + " '; ";

Reviewer: Add "// VULNERABLE CODE" here too? as this is the real line you don't want people to blindly copy and paste

```suggestion
    String QueryString = "select * from authors where lastname = ' " + search_lastname + " '; "; // VULNERABLE CODE
```

    rs = statement.executeQuery(QueryString); // VULNERABLE CODE
~~~~

> The intent is clear; if **search_lastname** has the value **Fred**, then the database will receive the query “**select * from authors where lastname='Fred';**” - a reasonable SQL query. But remember our warning signs - this code concatenates strings, some of that data is probably provided by an attacker, and we’re doing something called “execute”.  The warning signs are right. Imagine that the attacker provides the input “**Fred’ OR ’a’=’a**”. This will produce the query “**select * from authors where lastname='Fred' OR ’a’=’a’;**” and now the attacker can retrieve the entire database. The attacker could even modify or delete data this way, depending on various factors. This is a simple example of an SQL injection attack; an attacker can insert some characters and inject new or modified commands.

Reviewer: rm curly quotes

```suggestion
The intent is clear; if **search_lastname** has the value **Fred**, then the database will receive the query “**select * from authors where lastname='Fred';**” - a reasonable SQL query. But remember our warning signs - this code concatenates strings, some of that data is probably provided by an attacker, and we’re doing something called “execute”.  The warning signs are right. Imagine that the attacker provides the input “**Fred' OR 'a'='a**”. This will produce the query “**select * from authors where lastname='Fred' OR 'a'='a';**” and now the attacker can retrieve the entire database. The attacker could even modify or delete data this way, depending on various factors. This is a simple example of an SQL injection attack; an attacker can insert some characters and inject new or modified commands.
```

SQL injection vulnerabilities are one of the most common and devastating vulnerabilities, especially in web applications. They are also easy to counter, once you know how to do it.

> *Parameterized statements*, aka *parameterized queries*, are perhaps best way to counter SQL injection attacks if you are directly creating SQL commands that need to be secure. Parameterized statements let you identify placeholders (often a “**?**”) for data that needs to be escaped. A pre-existing library that you call then escapes the data properly for that specific implementation. The exact syntax for placeholders depends on the library and/or database you're using.

Reviewer: ```suggestion
*Parameterized statements*, aka *parameterized queries*, are perhaps the best way to counter SQL injection attacks if you are directly creating SQL commands that need to be secure. Parameterized statements let you identify placeholders (often a “**?**”) for data that needs to be escaped. A pre-existing library that you call then escapes the data properly for that specific implementation. The exact syntax for placeholders depends on the library and/or database you're using.
```

> For our purposes, a *prepared statement* prepares the statement with the database system ahead-of-time so that a later request can be executed more efficiently. Preparing a statement with a database ahead-of-time can help performance if the statement will be executed multiple times. Prepared statement APIs generally include support for parameterized statements, and many people (and APIs) use the terms "prepared statement" and "parameterized statement" as synonyms.

Reviewer: We don't define what "prepares the statement" means.

Author: Fair. "Compile" is clearer.

* `PQprepare` implements a prepared statement.
  It takes a statement with placeholders and
  submits it to the database to be prepared.
  Users can later use the separate `PQexecPrepared` call
  to provide the placeholder parameter values and execute the resulting
  command.

The [OWASP Query Parameterization Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Query_Parameterization_Cheat_Sheet.html) and [Bobby Tables website](https://bobby-tables.com/) provide examples for a variety of ecosystems.

##### Stored Procedures

Many database systems support "stored procedures", that is,
procedures embedded in the database itself.
If you use stored procedures, and commands (such as queries)
are dynamically constructed in them, then you can again have
SQL injection vulnerabilities.

Again, the best solution is usually to use a parameterized query mechanism
when creating the dynamic command (e.g., the SQL) inside the
stored procedure. Be careful; in some systems using them correctly
can be a little tricky.

For more information on using parameterized queries
in stored procedures, see your library's documentation, the

> [OWASP Query Parameterization Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Query_Parameterization_Cheat_Sheet.html), and the

Reviewer: link to specific section around stored procedures

```suggestion
[OWASP Query Parameterization Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Query_Parameterization_Cheat_Sheet.html#stored-procedure-examples), and the
```

[OWASP SQL Injection Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/SQL_Injection_Prevention_Cheat_Sheet.html)

> ##### When Parameterized Interfaces Won't Work

Reviewer: What are Parameterized Interfaces? We haven't defined them.

Author: Let's just use "parameterized statements" for consistency.

they do not allow varying information such as the names of tables, the names
of columns, or the sort order direction.

In those cases you may need to create a query by concatenating data.
In those cases, make *sure* you carefully validate the data (using an allowlist)
so only specific and safe values are allowed.
Often the allowlist is a short list of permitted values, or at most
a simple expression that only allows ASCII letters and digits.
A risk with this approach is that if the validation
is ever skipped (e.g., after some code change),
the system may be become extremely vulnerable.

The [OWASP SQL Injection Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/SQL_Injection_Prevention_Cheat_Sheet.html)
has a nice suggestion: if you *must* do this concatenation,
make the user input *different* from what you will actually use, and
in the program map the user input to the values you will use.
If user input doesn't have a valid mapping, reject the input.
This increases the probability that a change to the program
will not result in unvalidated input being concatenated into a query,
or that the problem will be detected before shipping.

> 

Reviewer: This is important and common enough that I think we should give a concrete example -- feel free to change:

```suggestion
For example, in Python, if you need to write to a user-provided table name, you can do the following:

~~~~python
    table_name_untrusted = request.get("table_name")  # This is untrusted, don't put this directly in the query!
    table_name_map = {"table1": "db.table1", "table2": "db.table2"}
    table_name = table_name_map[table_name_untrusted]
    con = sqlite3.connect(...)
    cur = con.cursor()
    cur.execute(f"insert into {table_name}(d, ts) values (?, ?)", (today, now)) # This is safe because we know that table_name can only take trusted values from table_name_map
~~~~
```

##### Other Approaches

Many programs use object-relational mapping (ORM). This is just a technique to automatically convert data in a relational database into an object in an object-oriented programming language and back; lots of libraries and frameworks will do this for you. This is fine, as long as the ORM is implemented using parameterized statements or something equivalent to them. In practice, any good ORM implementation will do so. So if you are using a respected ORM, you are already doing this. That said, it is common in systems that use ORMs to occasionally need to use SQL queries directly… and when you do, use parameterized statements or prepared statements.

> Some applications use a "query builder" library to build commands (queries) programmatically through a sequence of calls instead of embedding a command string. Some ORMs include a query builder system. Again, a well-implemented query builder will use parameterized statements or similar internally. So if you use a query builder, use that is implemented using parameterized statements and provide untrusted data as separate parameters however in the query builder requires for them to be correctly escaped.

Reviewer: ```suggestion
Some applications use a "query builder" library to build commands (queries) programmatically through a sequence of calls instead of embedding a command string. Some ORMs include a query builder system. Again, a well-implemented query builder will use parameterized statements or similar internally. So if you use a query builder, use one that is implemented using parameterized statements and provides untrusted data as separate parameters however in the query builder requires for them to be correctly escaped.
```

I also think this needs a grammar check ("provides untrusted data as separate parameters however in the query builder requires for them to be correctly escaped."), not sure what you meant to say

The [OWASP Query Parameterization Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Query_Parameterization_Cheat_Sheet.html) and [Bobby Tables website](https://bobby-tables.com/) provide examples for a variety of ecosystems.

##### Subtle issues: DBMS (Server) side vs. Application (client) side

An important security issue is *where* the
parameters of a parameterized statement are processed.
There are two options, DBMS-side and application-side, and
DBMS-side is better from a security point of view.

From a security point-of-view it's best if the parameters of

> parameterized statements are be processed directly

Reviewer: ```suggestion
parameterized statements are processed directly
```

within the database management system (DBMS),
aka "DBMS-side" parameter processing.
This approach is often called "server-side" since many DBMSs use
client/server architectures connecting over a network.
There are many advantages to DBMS-side parameter processing.
The DBMS has the current information on escaping rules

> (and can often use more efficient mechanisms than adding escape characters),

Reviewer: More efficient or also more effective?

Author: Efficient. "Effective" is covered by the text before the parentheses.

and it also has other information such as all relevant character
encodings.
Perhaps most importantly, the DBMS developers will typically have
security experts review this part of the DBMS system.
However, DBMS-side parameter processing often requires using special

> effort, so many DBMS libraries use "application-side" parameter processing.

Reviewer: ```suggestion
effort on the DBMS end, so many DBMS libraries use "application-side" parameter processing.
```

Author: It's not that it requires special effort on the DBMS side, it's that it requires added effort by the library author. That does point out a lack of clarity. I'll try to address that in a different way.

DBMS-side is better from a security point of view.

From a security point-of-view it's best if the parameters of
parameterized statements are processed directly
within the database management system (DBMS),
aka "DBMS-side" parameter processing.
This approach is often called "server-side" since many DBMSs use a
client/server architecture where the client connect over a network
to a server-side DBMS.
There are many advantages to DBMS-side parameter processing.
The DBMS has the current information on escaping rules
(and can often use more efficient mechanisms than adding escape characters),
and it also has other information such the relevant character encodings
and expected data types.
Perhaps most importantly, the DBMS developers will typically have
security experts review this part of the DBMS system.
However, DBMS-side parameter processing can require more effort to

> within implement in DBMS libraries, so many DBMS libraries use

Reviewer: ```suggestion implement within DBMS libraries, so many DBMS libraries use
```

   or situations that need to be escaped.
2. The application-side library may interpret multi-byte characters
   differently or not escape multi-byte characters correctly for the
   circumstance that actually exists on the DBMS side. (See
   [Multibyte character exploits PHP/MySQL](https://security.stackexchange.com/questions/9908/multibyte-character-exploits-php-mysql).)
3. If the application-side library implements parameterization of
   data types more complex than numbers and strings (such as arrays,
   objects, associative arrays, and/or dictionaries), then there
   is a significant risk of a vulnerability.
   The fundamental problem is that the application-side library isn't
   parsing the query language the same way that the DBMS would -
   it is doing simple text substitutions. So if the library implements this
   functionality, it must typically make *guesses* of what types are expected.

>    For example, it may guess that associative arrays are only provided sent

Reviewer: ```suggestion
   For example, it may guess that associative arrays are only sent
```

This increases the probability that a change to the program
will not result in unvalidated input being concatenated into a query,
or that the problem will be detected before shipping.
For example, in Python, if you need to write to a user-provided table name, you can do the following:

~~~~python
    table_name_untrusted = request.get("table_name")  # This is untrusted, don't put this directly in the query!
    table_name_map = {"table1": "db.table1", "table2": "db.table2"}
    table_name = table_name_map[table_name_untrusted]
    con = sqlite3.connect(...)
    cur = con.cursor()
    cur.execute(f"insert into {table_name}(d, ts) values (?, ?)", (today, now)) # This is safe because we know that table_name can only take trusted values from table_name_map

> ~~~~

Reviewer: ```suggestion
~~~~

```

Author: Will do!
